types/katex.d.ts

```diff
@@ -20,6 +20,43 @@ export type TrustContext =
     | { command: "\\htmlStyle", style: string }
     | { command: "\\htmlData", attributes: Record<string, string> }
 
+
+type KatexCatcodes = Record<string, number>;
+
+interface KatexLexerInterFace {
+  input: string;
+  tokenRegex: RegExp;
+  settings: Required<KatexOptions>;
+  catcodes: KatexCatcodes;
+}
+
+interface KatexSourceLocation {
+  start: number;
+  end: number;
+  lexer: KatexLexerInterFace;
+}
+
+export interface KatexToken {
+  text: string;
+  loc: KatexSourceLocation;
+  noexpand: boolean | undefined;
+  treatAsRelax: boolean | undefined;
+}
+
+
+export type KatexLoggerFunction = (
+  errorCode:
+    | "unknownSymbol"
+    | "unicodeTextInMathMode"
+    | "mathVsTextUnits"
+    | "commentAtEnd"
+    | "htmlExtension"
+    | "newLineInDisplayMode",
+  errorMsg: string,
+  token: KatexToken,
+) => boolean | "error" | "warn" | "ignore" | undefined | null | void;
```

Comment:
Actually I think `boolean | "error" | "warn" | "ignore" | void` is enough

Reply:
I don't think `| void` means what you think it means... I definitely wouldn't write that.

I agree that the docs don't allow `undefined | null` but the code explicitly says that it does. I could go either way on including them; probably `boolean` is enough. Or adding just `undefined` is maybe helpful.